backend/icloud/icloud.go

```diff
@@ -0,0 +1,1152 @@
+		_, resp, err = f.service.MoveItemToTrashByItemID(ctx, directoryID, etag, true)
+		return retryResultUnknown(ctx, resp, err)
+	}); err != nil {
+		return err
+	}
+
+	// flush everything from the left of the dir
+	f.dirCache.FlushDir(dir)
+
+	return nil
+}
+
+// Purge all files in the directory specified
+//
+// Implement this if you have a way of deleting all the files
+// quicker than just running Remove() on the result of List()
+//
+// Return an error if it doesn't exist
+func (f *Fs) Purge(ctx context.Context, dir string) error {
+	if dir == "" {
+		return fs.ErrorCantPurge
+	}
+	return f.purgeCheck(ctx, dir, false)
+}
+
+func (f *Fs) listAll(ctx context.Context, dirID string) (items []*api.DriveItem, err error) {
+	service, _ := f.icloud.DriveService()
+	var itemsRaw []*api.DriveItemRaw
+	var resp *http.Response
+	if err = f.pacer.Call(func() (bool, error) {
+		id, _ := f.parseNormalizedID(dirID)
+		itemsRaw, resp, err = service.GetItemsInFolder(ctx, id, 100000)
+		return shouldRetry(ctx, resp, err)
+	}); err != nil {
+		return nil, err
+	}
+
+	// var driveItems []*api.DriveItemRaw
+	for _, i := range itemsRaw {
+		item := i.IntoDriveItem()
+		item.Name = f.opt.Enc.ToStandardName(item.Name)
+		item.Extension = f.opt.Enc.ToStandardName(item.Extension)
+		items = append(items, item)
+	}
+	//fs.PrettyPrint(items, "ERR", fs.LogLevelInfo)
+
+	return items, nil
+}
+
+// List implements fs.Fs.
+func (f *Fs) List(ctx context.Context, dir string) (entries fs.DirEntries, err error) {
+	dirRemoteID, err := f.dirCache.FindDir(ctx, dir, false)
+	if err != nil {
+		return nil, err
+	}
+
+	entries = make(fs.DirEntries, 0)
+	items, _ := f.listAll(ctx, dirRemoteID)
+
+	for _, item := range items {
+		id := item.Itemid
+		name := item.FullName()
+		remote := path.Join(dir, name)
+		if item.IsFolder() {
+			jid := f.putFolderCache(id, item.Etag, remote)
+			d := fs.NewDir(remote, item.DateModified).SetID(jid)
+			entries = append(entries, d)
+		} else {
+			o, _ := f.NewObjectFromDriveItem(ctx, remote, item)
+			entries = append(entries, o)
+		}
+	}
+
+	return entries, nil
+}
+
+// Mkdir implements fs.Fs.
+func (f *Fs) Mkdir(ctx context.Context, dir string) error {
+	// _, err := f.dirCache.FindDir(ctx, dir, true)
+	_, _, err := f.FindDir(ctx, dir, true)
+	return err
+}
+
+// Name implements fs.Fs.
+func (f *Fs) Name() string {
+	// fs.Debugf("Name:", f.name)
+	return f.name
+}
+
+// Precision implements fs.Fs.
+func (f *Fs) Precision() time.Duration {
+	return time.Second
+}
+
+// Copy src to this remote using server-side copy operations.
+//
+// This is stored with the remote path given.
+//
+// It returns the destination Object and a possible error.
+//
+// Will only be called if src.Fs().Name() == f.Name()
+//
+// If it isn't possible then return fs.ErrorCantCopy
+func (f *Fs) Copy(ctx context.Context, src fs.Object, remote string) (fs.Object, error) {
+	// note: so many calls its only just faster then a reupload for big files.
```

Comment:
```suggestion
	// Note: so many calls its only just faster then a reupload for big files.
```

That setence sounds a bit strange.